backend/drive/drive_activity.go

```diff
@@ -0,0 +1,240 @@
+		if action.Detail.Edit != nil {
+			fs.Debugf(targetTitle, "Edited")
+			invalidatedParent = f.getParent(target)
+			invalidatedParents = addParent(invalidatedParents, invalidatedParent)
+		}
+		*/
+		/* TODO: Verify if this is needed
+		if action.Detail.PermissionChange != nil {
+			fs.Debugf(targetTitle, "Permissions Changed")
+			invalidatedParent = f.getParent(target)
+			invalidatedParents = addParent(invalidatedParents, invalidatedParent)
+		}
+		*/
+		if action.Detail.Move != nil {
+			fs.Debugf(targetTitle, "Moved")
+			moveAction := action.Detail.Move
+			for _, addedParent := range moveAction.AddedParents {
+				if addedParent.DriveItem == nil {
+					fs.Errorf(targetTitle, "Invalid Parent: %v\n", addedParent)
+				}
+				parentItem := addedParent.DriveItem
+				invalidatedParent = &parent{
+					ID: parseItemID(parentItem.Name), Name: parentItem.Title}
+				invalidatedParents = addParent(invalidatedParents, invalidatedParent)
+			}
+			for _, removedParent := range moveAction.RemovedParents {
+				if removedParent.DriveItem == nil {
+					fs.Errorf(targetTitle, "Invalid Parent: %v\n", removedParent)
+				}
+				parentItem := removedParent.DriveItem
+				invalidatedParent = &parent{
+					ID: parseItemID(parentItem.Name), Name: parentItem.Title}
+				invalidatedParents = addParent(invalidatedParents, invalidatedParent)
+			}
+		}
+		for _, invalidatedParent := range invalidatedParents {
+			fs.Debugf(targetTitle, "Invalidated Parent : %s", invalidatedParent)
+		}
+	}
+	return
+}
+
+func (f *Fs) parseActivity(activity *driveactivity.DriveActivity) map[string]string {
+	target := getTargetInfo(activity.Targets)
+	return f.parseActivityActions(target, activity.Actions)
+}
+
+func (f *Fs) changeNotifyRunner(ctx context.Context, fromTime,
+	toTime time.Time, notifyFunc func(string, fs.EntryType)) {
+
+	var (
+		err                error
+		pageToken          string
+		pathsToClear       []entryType
+		invalidatedParents map[string]string
+		activityQuery      *driveactivity.QueryDriveActivityRequest
+		activityResponse   *driveactivity.QueryDriveActivityResponse
+	)
+
+	filter := fmt.Sprintf(GlobalQuery,
+		fromTime.Format(time.RFC3339), toTime.Format(time.RFC3339))
+
+	for {
+		activityQuery = createGlobalQueryRequest(f.rootFolderID, filter, pageToken)
+
+		err = f.pacer.Call(func() (bool, error) {
+			activityResponse, err = f.activitySvc.Activity.Query(activityQuery).Context(ctx).Do()
+			return f.shouldRetry(err)
+		})
+		if err != nil {
+			fs.Errorf(f, "Unable to retrieve list of activities: %v", err)
+			return
+		}
+		fs.Infof(f, "Retrieved Activities : %d\n", len(activityResponse.Activities))
+
+		for _, activity := range activityResponse.Activities {
+			invalidatedParents = f.parseActivity(activity)
+
+			for parentID, parentName := range invalidatedParents {
+				parentName = f.opt.Enc.ToStandardName(parentName)
+				// translate the path of this dir
+				if parentPath, ok := f.dirCache.GetInv(parentID); ok {
```

Comment:
Looking things up in the dirCache only works if we have listed that directory in the past.

I can't immediately think of a better solution, but I thought I'd point it out.

The old code has this problem too.

Reply:
Is that actually a problem? I would have assumed that's perfectly fine since dirs which have not been listed don't need to be invalidated. It will anyway fetch the current listing from the remote when it goes for listing.